### PR TITLE
Add arduino service

### PR DESCRIPTION
**Pull request by gregersrygg**
Move the Arduino logic to go instead of bash. When starting the service it will compile the arduino sketch and upload if it compiled successfully. Then it will start logging the serial output to a log file.

### CODE REVIEW

**Comment by gregersrygg** [comment]
@gordonklaus Merging now. Feel free to review anyways :)

File: arduino-service/arduino.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+type Arduino struct {
+	board, port string
+	log         *log.Logger
+}
+
+func NewArduino(board, port string, logName string) *Arduino {
+	logFile, err := os.OpenFile(logName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		log.Println("Error: ", err)
+	}
```
Comment (gordonklaus) [inline arduino-service/arduino.go]: `return nil, err`?

File: arduino-service/arduino.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+func (a *Arduino) MonitorSerial() {
+	a.log.Printf("Starting serial connection to %s (%s)\n", a.port, a.board)
+	serialConfig := &serial.Config{Name: a.port, Baud: 9600, ReadTimeout: time.Second * 1}
+
+	serial, err := serial.OpenPort(serialConfig)
+	if err != nil {
+		a.log.Println("Error: ", err)
```
Comment (gordonklaus) [inline arduino-service/arduino.go]: retry?

File: arduino-service/arduino.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	a.log.Println("Verifying " + a.board)
+	return arduino(a.log, "compile", "--fqbn", a.board, sketch)
+}
+
+func (a *Arduino) Upload(sketch string) error {
+	a.log.Println("Uploading to " + a.port)
+	return arduino(a.log, "upload", "-p", a.port, "--fqbn", a.board, sketch)
+}
+
+func dropCR(data []byte) []byte {
```
Comment (gordonklaus) [inline arduino-service/arduino.go]: Could use https://golang.org/pkg/bytes/#TrimRight

File: arduino-service/arduino.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	return 0, nil, nil
+}
+
+func arduino(log *log.Logger, args ...string) error {
+	return exe(log, "arduino-cli", args...)
+}
+
+func exe(log *log.Logger, name string, args ...string) error {
+	cmd := exec.Command(name, args...)
+	stdout, err := cmd.StdoutPipe()
```
Comment (gordonklaus) [inline arduino-service/arduino.go]: sure there's no stderr output?

File: arduino-service/main.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+		if err := a.Verify(opts.Sketch); err != nil {
+			log.Println("Error: ", err)
+		} else {
+			if err = a.Upload(opts.Sketch); err != nil {
+				log.Println("Error: ", err)
+			}
+		}
+
+		log.Printf("Starting serial monitor for %s %s\n", opts.Boards[0], opts.Ports[0])
```
Comment (gordonklaus) [inline arduino-service/main.go]: I think you meant `[i]` not `[0]`.  Maybe factor out `port := opts.Ports[i]` which is also used above.

File: arduino-service/main.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+		log.Fatalf("%s is not a directory", opts.LogDir)
+	}
+
+	if len(opts.Boards) > len(opts.Ports) {
+		log.Fatalln("Error: more boards than ports")
+	} else if len(opts.Ports) > len(opts.Boards) {
+		log.Fatalln("Error: more ports than boards")
+	}
+
+	for i, b := range opts.Boards {
```
Comment (gordonklaus) [inline arduino-service/main.go]: rename `b` to `board` for clarity

File: arduino-service/arduino.go (hunk 1 above)
Comment (gregersrygg) [reply arduino-service/arduino.go]: why not just a `log.Fatalln` directly?

Comment (gordonklaus) [reply arduino-service/arduino.go]: sure, why not.

**Pull request merged**